meshnet/monitor.go

```diff
@@ -0,0 +1,99 @@
+package meshnet
+
+import (
+	"log"
+	"sync"
+
+	"github.com/NordSecurity/nordvpn-linux/internal"
+	"github.com/vishvananda/netlink"
+)
+
+type (
+	PID     uint32
+	SetupFn func() (MonitorChannels, error)
+)
+
+type MonitorChannels struct {
+	EventCh chan netlink.ProcEvent
+	DoneCh  chan struct{}
+	ErrCh   chan error
+}
+
+type ProcEvent struct {
+	PID PID
+}
+
+// EventHandler processes events of process started and stopped.
+type EventHandler interface {
+	OnProcessStarted(ProcEvent)
+	OnProcessStopped(ProcEvent)
+}
+
+// NetlinkProcessMonitor monitors EXEC and EXIT events of processes and calls
+// [EventHandler.OnProcessStarted] and [EventHandler.OnProcessStopped] accordingly.
+type NetlinkProcessMonitor struct {
+	handler EventHandler
+	setup   SetupFn
+}
+
+func NewProcMonitor(handler EventHandler, setup SetupFn) NetlinkProcessMonitor {
+	return NetlinkProcessMonitor{
+		handler: handler,
+		setup:   setup,
+	}
+}
+
+// Start starts listening for process events using Process Events Connector netlink API.
+//
+// It recreates the source of the events by calling [SetupFn]
+// every time [NetlinkProcessMonitor.Start] is called.
+func (pm *NetlinkProcessMonitor) Start() (MonitorStopper, error) {
+	channels, err := pm.setup()
+	if err != nil {
+		return MonitorStopper{}, err
+	}
+
+	stopCh := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case ev := <-channels.EventCh:
+				pm.handleProcessEvent(&ev)
+			case <-stopCh:
```

Comment:
Maybe context would be useful in this case? I.e `Start` takes a context instead of returning `MonitorStopper` and passes it to `setup` function.

Reply:
Yes, it's better with context :+1:, updated